--- lib-python/2.7/distutils/command/install.py	2015-04-18 05:34:44.619111250 +0300
+++ Lib/distutils/command/install.py	2015-04-18 05:34:31.767089732 +0300
@@ -83,6 +83,20 @@
         'scripts': '$userbase/bin',
         'data'   : '$userbase',
         },
+    'java': {
+        'purelib': '$base/Lib/site-packages',
+        'platlib': '$base/Lib/site-packages',
+        'headers': '$base/Include/$dist_name',
+        'scripts': '$base/bin',
+        'data'   : '$base',
+        },
+    'java_user': {
+        'purelib': '$usersite',
+        'platlib': '$usersite',
+        'headers': '$userbase/include/python$py_version_short/$dist_name',
+        'scripts': '$userbase/bin',
+        'data'   : '$userbase',
+        },
     }
 
 # The keys to an installation scheme; if any new types of files are to be
