--- lib-python/2.7/doctest.py	2015-04-18 05:34:44.687111364 +0300
+++ Lib/doctest.py	2015-04-18 05:34:31.779089753 +0300
@@ -1022,6 +1022,8 @@
             filename = getattr(module, '__file__', module.__name__)
             if filename[-4:] in (".pyc", ".pyo"):
                 filename = filename[:-1]
+            elif filename.endswith('$py.class'):
+                filename = '%s.py' % filename[:-9]
         return self._parser.get_doctest(docstring, globs, name,
                                         filename, lineno)
 
@@ -1738,7 +1740,8 @@
 
        If a failure or error occurs, the globals are left intact:
 
-         >>> del test.globs['__builtins__']
+         >>> if '__builtins__' in test.globs:
+         ...     del test.globs['__builtins__']
          >>> test.globs
          {'x': 1}
 
@@ -1752,7 +1755,8 @@
          ...
          UnexpectedException: <DocTest foo from foo.py:0 (2 examples)>
 
-         >>> del test.globs['__builtins__']
+         >>> if '__builtins__' in test.globs:
+         ...     del test.globs['__builtins__']
          >>> test.globs
          {'x': 2}
 
@@ -2398,6 +2402,8 @@
             filename = module.__file__
             if filename[-4:] in (".pyc", ".pyo"):
                 filename = filename[:-1]
+            elif filename.endswith('$py.class'):
+                filename = '%s.py' % filename[:-9]
             test.filename = filename
         suite.addTest(DocTestCase(test, **options))
 
