--- lib-python/2.7/urllib.py	2015-04-18 05:34:44.823111590 +0300
+++ Lib/urllib.py	2015-04-18 05:34:31.883089926 +0300
@@ -44,7 +44,7 @@
 MAXFTPCACHE = 10        # Trim the ftp cache beyond this size
 
 # Helper for non-unix systems
-if os.name == 'nt':
+if (os._name if sys.platform.startswith('java') else os.name) == 'nt':
     from nturl2path import url2pathname, pathname2url
 elif os.name == 'riscos':
     from rourl2path import url2pathname, pathname2url
