--- lib-python/2.7/xml/dom/xmlbuilder.py	2015-04-18 05:34:44.843111625 +0300
+++ Lib/xml/dom/xmlbuilder.py	2015-04-18 05:34:31.887089934 +0300
@@ -3,6 +3,8 @@
 import copy
 import xml.dom
 
+from xml.dom.minicompat import *
+
 from xml.dom.NodeFilter import NodeFilter
 
 
@@ -79,7 +81,7 @@
                 settings = self._settings[(_name_xform(name), state)]
             except KeyError:
                 raise xml.dom.NotSupportedErr(
-                    "unsupported feature: %r" % (name,))
+                    "unsupported feature: " + `name`)
             else:
                 for name, value in settings:
                     setattr(self._options, name, value)
@@ -91,7 +93,7 @@
 
     def canSetFeature(self, name, state):
         key = (_name_xform(name), state and 1 or 0)
-        return key in self._settings
+        return self._settings.has_key(key)
 
     # This dictionary maps from (feature,value) to a list of
     # (option,value) pairs that should be set on the Options object.
@@ -209,7 +211,7 @@
     return name.lower().replace('-', '_')
 
 
-class DOMEntityResolver(object):
+class DOMEntityResolver(NewStyle):
     __slots__ = '_opener',
 
     def resolveEntity(self, publicId, systemId):
@@ -247,13 +249,13 @@
 
     def _guess_media_encoding(self, source):
         info = source.byteStream.info()
-        if "Content-Type" in info:
+        if info.has_key("Content-Type"):
             for param in info.getplist():
                 if param.startswith("charset="):
                     return param.split("=", 1)[1].lower()
 
 
-class DOMInputSource(object):
+class DOMInputSource(NewStyle):
     __slots__ = ('byteStream', 'characterStream', 'stringData',
                  'encoding', 'publicId', 'systemId', 'baseURI')
 
