--- lib-python/2.7/test/test_defaultdict.py	2015-04-18 05:34:44.511111069 +0300
+++ Lib/test/test_defaultdict.py	2015-04-18 05:34:31.827089833 +0300
@@ -148,6 +148,7 @@
         else:
             self.fail("expected KeyError")
 
+    @unittest.skipIf(test_support.is_jython, "FIXME: incorrect formatting of default_factory when it's a bound method")
     def test_recursive_repr(self):
         # Issue2045: stack overflow when default_factory is a bound method
         class sub(defaultdict):
