--- lib-python/2.7/test/test_glob.py	2015-04-18 05:34:44.567111162 +0300
+++ Lib/test/test_glob.py	2015-04-18 05:34:31.835089846 +0300
@@ -69,9 +69,15 @@
         eq(self.glob('zymurgy'), [])
 
         res = glob.glob('*')
-        self.assertEqual({type(r) for r in res}, {str})
+        # For a clean checkout, the next two assertions would never
+        # have failed, even with the change with Jython in
+        # https://hg.python.org/jython/rev/ea036792f304
+        #
+        # But for developers playing with things, we should not have
+        # it fail either
+        self.assertLessEqual({type(r) for r in res}, {str, unicode})
         res = glob.glob(os.path.join(os.curdir, '*'))
-        self.assertEqual({type(r) for r in res}, {str})
+        self.assertLessEqual({type(r) for r in res}, {str, unicode})
 
         # test return types are unicode, but only if os.listdir
         # returns unicode filenames
